lib/worker/exec.go

```diff
@@ -0,0 +1,129 @@
+package worker
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+// Exec represents something to be executed in the Worker
+// TODO: I might just remove this interface in the next
+// PR. I introduced it thinking it could be useful for
+// the future cgroup implementation but now am not so
+// sure.
+type Exec interface {
+	// Command is the command for the exec.
+	Command() string
+	// Start starts the exec.
+	Start() error
+	// Stop stops the exec.
+	Stop() error
+	// PID returns the pid of the process once it has been started.
+	// Otherwise, and invalid pid 0 is returned.
+	PID() int
+	// Status returns the ProcessStatus struct corresponding
+	// to the started exec.
+	Status() (*ProcessStatus, error)
+}
+
+// process is a process executed or to be executed by the worker
+// TODO: Find a better name for this as the word "process" implies
+// that this it is already running.
+type process struct {
+	command    string
+	cmd        *exec.Cmd
+	startedBy  string
+	startedAt  time.Time
+	finishedAt time.Time // TODO: implement finished at
+}
+
+func newProcess(command string, cmd *exec.Cmd, startedBy string) *process {
+	return &process{
+		command:   command,
+		cmd:       cmd,
+		startedBy: startedBy,
+	}
+}
+
+// ProcessStatus represents a snapshot of a process status
+type ProcessStatus struct {
+	PID        int
+	StartedBy  string
+	State      string
+	StartedAt  time.Time
+	FinishedAt time.Time // TODO: implement finished at
+}
+
+func (p *process) Start() error {
+	p.startedAt = time.Now()
```

Comment:
What happens if  `Start()` is called multiple times?

Reply:
It would fail as the underlying `cmd.Start()` fails with `exec: already started` if an attempt to start it twice is made.

Reply:
If it is called at the right time it might execute twice. Was sort of relying on the worker doing reasonable things since this is a private type but that is definitely not a good strategy. I will make the Start and Stop methods mutually exclusive.

PS: all these methods should be private, will change this.

Reply:
Done, also move the process type to its own packaged and removed the Exec interface as it wasn't really useful right now.